src/utils/async.ts

```diff
@@ -30,6 +30,41 @@ export class ManualPromise {
     }
 }
 
+export class WaitGroup {
```

Comment:
needs a docstring that calls out the purpose and behavior. seems a bit weird that `add()` takes no arguments